### PR TITLE
Handle inactive enum variants in `MaybeUninitializedPlaces`

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Resolves the first part of #69715.

This is the equivalent of #68528 but for `MaybeUninitializedPlaces`. Because we now notify drop elaboration that inactive enum variants might be uninitialized, some drops get marked as ["open" that were previously "static"](https://github.com/rust-lang/rust/blob/e0e5d82e1677c82d209b214bbfc2cc5705c2336a/src/librustc_mir/transform/elaborate_drops.rs#L191). Unlike in #69715, this isn't strictly better: An "open" drop expands to more MIR than a simple call to the drop shim. However, because drop elaboration considers each field of an "open" drop separately, it can sometimes eliminate unnecessary drops of moved-from or unit-like enum variants. This is the case for `Option::unwrap`, which is reflected in the `mir-opt` test.

cc @eddyb 
r? @oli-obk

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
@bors try
@rust-timer queue

**Comment by Aaron1011** [comment]
@ecstatic-morse: How does interact with generators?

**Comment by jonas-schievink** [comment]
Neat!

**Comment by ecstatic-morse** [comment]
> How does interact with generators?

Hey, you're not a :ghost: :smile:!

The generator transform doesn't use `MaybeUninitializedLocals` to compute generator saved locals, and it does drop elaboration with `DropShimElaborator`                                                                , which also doesn't use `MaybeUninitializedLocals`. Only `Elaborator`, which seems to be the "general purpose" drop elaborator uses it. Furthermore, `apply_discriminant_switch_effect` doesn't get called for `TyKind::Generator`s, so I don't believe that this affects generators any more than regular functions. I don't know what the benefits might be from changing that last part.

**Comment by ecstatic-morse** [comment]
r? @oli-obk 
cc @eddyb

File: src/test/mir-opt/simplify-locals-removes-unused-discriminant-reads/32bit/rustc.map.SimplifyLocals.diff [hunk 1]
```diff
@@ -7,13 +7,28 @@
       let mut _2: isize;                   // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:3:9: 3:13
       let _3: std::boxed::Box<()>;         // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:4:14: 4:15
 -     let mut _4: std::boxed::Box<()>;     // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:4:25: 4:26
--     let mut _5: isize;                   // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:6:1: 6:2
+-     let mut _5: bool;                    // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:6:1: 6:2
 -     let mut _6: isize;                   // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:6:1: 6:2
+-     let mut _7: isize;                   // in scope 0 at $DIR/simplify-locals-removes-unused-discriminant-reads.rs:6:1: 6:2
```
Comment (oli-obk) [inline src/test/mir-opt/simplify-locals-removes-unused-discriminant-reads/32bit/rustc.map.SimplifyLocals.diff:12]: Oh god diffs of diffs XD. I love how this basically says we now have an additional variable, but it's optimized out here.

**Comment by oli-obk** [comment]
You're really raising the bar on what a high quality PR means... Readable PR diff, modular design, documentation and it all causes a perf improvement along the way.

**Comment by JohnTitor** [comment]
Another case of #73992, @bors retry

**Pull request merged**